### PR TITLE
Onboard FLUX model step 1: Adding dataloader, model code and forward pass

### PR DESCRIPTION
**Pull request by wwwjn**
## Context
Onboard FLUX model to torchtitan , FLUX implementation: https://github.com/black-forest-labs/flux/

## Run
Test dataloader only:
`pytest tests/unit_tests/test_flux_dataloader.py -s`

Run forward pass:
`PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True torchrun --nproc_per_node=1 torchtitan/experiments/flux/train.py --job.config_file torchtitan/experiments/flux/train_configs/debug_model.toml`
Screenshot 2025-04-04 at 3 59 14 PM
- Current loss is high because we haven't implement `init_weights()` function in `model.py`

## Next Step
1. Add FSDP to flux model

### CODE REVIEW

File: torchtitan/datasets/flux_dataset.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (fegin) [inline torchtitan/datasets/flux_dataset.py]: should we put this in the `experiments/flux` as well?

File: torchtitan/experiments/flux/modules/HFEmbedder.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+from torch import nn, Tensor
+from transformers import CLIPTextModel, CLIPTokenizer, T5EncoderModel, T5Tokenizer
+
+
+class HFEmbedder(nn.Module):
```
Comment (fegin) [inline torchtitan/experiments/flux/modules/HFEmbedder.py]: noob question, what does this HF stand for? hugging face?

Comment (wwwjn) [reply torchtitan/experiments/flux/modules/HFEmbedder.py]: The name was adopted from Flux repo, my understanding is HF=hugging_face. Then embedder is downloaded from hugging face :)

File: torchtitan/experiments/flux/modules/layers.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        super().__init__()
+        self.in_layer = nn.Linear(in_dim, hidden_dim, bias=True)
+        self.silu = nn.SiLU()
+        self.out_layer = nn.Linear(hidden_dim, hidden_dim, bias=True)
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.out_layer(self.silu(self.in_layer(x)))
+
+
+class RMSNorm(torch.nn.Module):
```
Comment (lessw2020) [inline torchtitan/experiments/flux/modules/layers.py]: I think you can just use PyTorch's (new) nn.RMSNorm here to be consistent with Titan:
nn.RMSNorm(dim, eps=eps)
see: 
https://github.com/pytorch/torchtitan/blob/3c1298518a3d46ae70156c48095dbae70d30dd15/torchtitan/models/norms.py#L33

Comment (wwwjn) [reply torchtitan/experiments/flux/modules/layers.py]: Sure, thank you! These part need to be further integrated into torchtitan

File: .ci/docker/requirements.txt [hunk 4]
```diff
@@ -7,3 +7,7 @@ blobfile
 tabulate
 wandb
 fsspec
+transformers
+einops
+sentencepiece
+torchvision
```
Comment (tianyu-l) [inline .ci/docker/requirements.txt]: let's put these into a separate file, only within the flux experiment folder

File: tests/unit_tests/test_flux_dataloader.py
Comment (tianyu-l) [inline tests/unit_tests/test_flux_dataloader.py:1]: Let's put this test files into flux folder, too. The idea is not to let core part be exposed to experiments content. We can set up dedicated CI workflow to call those tests.

File: torchtitan/datasets/flux_dataset.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/datasets/flux_dataset.py]: I agree. In the beginning, let's make `experiments/flux` hosting all flux-related files.

File: torchtitan/experiments/flux/modules/HFEmbedder.py
Comment (tianyu-l) [inline torchtitan/experiments/flux/modules/HFEmbedder.py:1]: file name style should be consistent with others, e.g. `hf_embedder.py`

File: torchtitan/experiments/flux/model.py
Comment (tianyu-l) [inline torchtitan/experiments/flux/model.py:1]: Let's create a folder `model` to host all model code files, including `model.py` and files in `modules`. Later we can create another folder `infra` to host all parallelisms related files, if there are more than one.

File: torchtitan/experiments/flux/model_builder.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+from torchtitan.experiments.flux.model import FluxModel, FluxParams
+from torchtitan.experiments.flux.modules.autoencoder import (
+    AutoEncoder,
+    AutoEncoderParams,
+)
+from torchtitan.experiments.flux.modules.embedder import HFEmbedder
+
+
+@dataclass
+class ModelSpec:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model_builder.py]: As discussed offline, let's work on making things consistent with torchtitan components and protocols, including `JobConfig`, `ModelArgs`, `TrainSpec`, etc.

Comment (wwwjn) [reply torchtitan/experiments/flux/model_builder.py]: Working on it :)

File: .ci/docker/requirements.txt (hunk 4 above)
Comment (tianyu-l) [inline .ci/docker/requirements.txt]: can we move these to Flux folder only?

File: torchtitan/experiments/flux/flux_argparser.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        "--encoder.clip_encoder",
+        type=str,
+        default="openai/clip-vit-large-patch14",
+        help="Clip encoder to use, hugging face model name.",
+    )
+    parser.add_argument(
+        "--encoder.encoder_dtype",
+        type=torch.dtype,
+        default=torch.bfloat16,
+        help="Where to load the encoder, cpu or cuda.",
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/flux_argparser.py]: why the dtype config helper message is about device?

File: torchtitan/experiments/flux/math.py
Comment (tianyu-l) [inline torchtitan/experiments/flux/math.py:1]: let's put this file under the `model` folder

File: torchtitan/experiments/flux/model/modules/autoencoder.py
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/modules/autoencoder.py:1]: I think we can put all files under `modules` just under `model`, i.e. one layer above

File: torchtitan/train.py [hunk 7]
```diff
@@ -12,10 +12,10 @@
 
 import torch
 
-from torch.distributed.elastic.multiprocessing.errors import record
-
 import torchtitan.components.ft as ft
 import torchtitan.protocols.train_spec as train_spec_module
+
+from torch.distributed.elastic.multiprocessing.errors import record
```
Comment (tianyu-l) [inline torchtitan/train.py]: do we need to move this?

File: torchtitan/components/tokenizer.py [hunk 8]
```diff
@@ -11,11 +11,12 @@
 
 class Tokenizer(ABC):
     # basic tokenizer interface, for typing purpose mainly
-    def __init__(self, tokenizer_path: str):
-        assert os.path.exists(
-            tokenizer_path
-        ), f"The tokenizer path does not exist: {tokenizer_path}"
-        assert os.path.isfile(tokenizer_path), tokenizer_path
+    def __init__(self, tokenizer_path: str, from_pretrained: bool = False):
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: instead of adding the `from_pretrained` arg, let's
- remove the `tokenizer_path` arg and its assertion below
- remove it also from the dependency here https://github.com/pytorch/torchtitan/blob/main/torchtitan/datasets/tokenizer/tiktoken.py#L38

File: torchtitan/experiments/flux/dataset/tokenizer.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    Build the tokenizer from the job config.
+    """
+    t5_tokenizer = FluxTokenizer(
+        model_path=job_config.encoder.t5_encoder,
+        max_length=job_config.encoder.max_t5_encoding_len,
+    )
+    return t5_tokenizer
+
+
+def build_flux_clip_tokenizer(job_config: JobConfig) -> FluxTokenizer:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/tokenizer.py]: this is not called? let's remove the two build tokenizer functions.

File: torchtitan/components/loss.py [hunk 10]
```diff
@@ -27,3 +27,16 @@ def build_cross_entropy_loss(job_config: JobConfig):
         logger.info("Compiling the loss function with torch.compile")
         loss_fn = torch.compile(loss_fn)
     return loss_fn
+
+
+def mse_loss(pred: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: Let's put this in flux folder for now, as it's not clear if other models will need it.
Later we can "upstream" it here if that's the case.

File: torchtitan/experiments/flux/README.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Flux model in torchtitan
+
+## Overview
+
+## Usage
+
+Run the following command to train the model:
+```
+    PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True torchrun --nproc_per_node=1 torchtitan/experiments/flux/train.py --job.config_file torchtitan/experiments/flux/train_configs/debug_model.toml
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/README.md:9]: please add a TODO that parallelism is going to be added soon

File: torchtitan/experiments/flux/__init__.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+__all__ = [
+    "FluxModelArgs",
+    "FluxModel",
+    "flux_configs",
+    "parallelize_flux",
+]
+
+
+flux_configs = {
+    "flux-dev": FluxModelArgs(
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/__init__.py:28]: are there multiple versions in the flux repo? what are the differences?
I think it'll be convenient if we just "make up" a very small model, by reducing layers, dims, etc., so that we can easily work with for development purpose. It could be in a separate PR though.

File: torchtitan/experiments/flux/dataset/flux_dataset.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+
+def _process_cc12m_image(
+    img: Image.Image,
+    img_height: int = 256,
+    img_width: int = 256,
+) -> Image.Image:
+    """Process CC12M image to the desired size."""
+
+    # width, height = img.size
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/flux_dataset.py]: what is the commented code for?

File: torchtitan/experiments/flux/__init__.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    TrainSpec(
+        name="flux",
+        cls=FluxModel,
+        config=flux_configs,
+        parallelize_fn=parallelize_flux,
+        pipelining_fn=None,
+        build_optimizers_fn=build_optimizers,
+        build_lr_schedulers_fn=build_lr_schedulers,
+        build_dataloader_fn=build_flux_dataloader,
+        build_tokenizer_fn=build_flux_t5_tokenizer,
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/__init__.py]: Because Flux is a bit special and uses two tokenizers simultaneously, this field is less helpful. Let's mark this field as optional in `train_spec.py`, and send None here.

File: torchtitan/experiments/flux/train.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        bsz = labels.shape[0]
+
+        with torch.no_grad():
+            noise = torch.randn_like(labels)
+            timesteps = torch.rand((bsz,)).to(labels)
+            sigmas = timesteps.view(-1, 1, 1, 1)
+            noisy_latents = (1 - sigmas) * labels + sigmas * noise
+            guidance = torch.full((bsz,), self._guidance).to(labels)
+
+        target = noise - labels
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train.py:148]: we should put these in the loss function

File: torchtitan/experiments/flux/utils.py
Comment (tianyu-l) [inline torchtitan/experiments/flux/utils.py:1]: It seems we are mixing multiple functions in `utils.py` and `sampling.py`. I think we need clearer logical organization of the code. If it's for training only, e.g. `predict_noise`, we probably should put it in `train.py`.

File: torchtitan/experiments/flux/__init__.py (hunk 12 above)
Comment (wwwjn) [reply torchtitan/experiments/flux/__init__.py:28]: Yes there are multiple versions in the flux repo, and they have different model parameters. I adopted one version "flux-dev" params but yes, this model is not a small model. I could make another PR to make a small debug model.

File: torchtitan/experiments/flux/train.py (hunk 15 above)
Comment (wwwjn) [reply torchtitan/experiments/flux/train.py:148]: As discussed offline, we will leave this part in train_step for simplicity.

File: torchtitan/experiments/flux/requirements.txt [hunk 16]
```diff
@@ -0,0 +1,3 @@
+transformers
+sentencepiece
+torchvision
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/requirements.txt]: can we remove this?

File: torchtitan/experiments/flux/requirements.txt [hunk 17]
```diff
@@ -0,0 +1,3 @@
+transformers
+sentencepiece
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/requirements.txt]: where is this used?

File: torchtitan/experiments/flux/dataset/tokenizer.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        return tokens
+
+    def decode(self, t: List[int]) -> str:
+        """
+        Decode function. This function will not be called.
+        """
+        return self._tokenizer.decode(t)
+
+
+def build_flux_t5_tokenizer(job_config: JobConfig) -> FluxTokenizer:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/dataset/tokenizer.py]: please remove if not called

File: torchtitan/experiments/flux/__init__.py (hunk 12 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/__init__.py:28]: Let's add `flux-dev`, `flux-schnell`, and a debug model

File: torchtitan/experiments/flux/train_configs/debug_model.toml [hunk 19]
```diff
@@ -0,0 +1,68 @@
+
+[job]
+dump_folder = "./outputs"
+description = "Flux1.1-dev model"
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/train_configs/debug_model.toml]: modify this

File: torchtitan/datasets/tokenizer/tiktoken.py [hunk 20]
```diff
@@ -35,7 +35,7 @@ class TikTokenizer(Tokenizer):
     pat_str = r"(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"  # noqa: E501, B950
 
     def __init__(self, model_path: str):
-        super().__init__(model_path)
+        super().__init__()
         assert os.path.isfile(model_path), model_path
```
Comment (tianyu-l) [inline torchtitan/datasets/tokenizer/tiktoken.py:42]: could add the check
```
assert os.path.exists(
    tokenizer_path
), f"The tokenizer path does not exist: {tokenizer_path}"
```

File: torchtitan/experiments/flux/model/model.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+    guidance_embed: bool = True
+    autoencoder_params: AutoEncoderParams = field(default_factory=AutoEncoderParams)
+
+    def update_from_config(self, job_config: JobConfig, tokenizer: Tokenizer) -> None:
+        # context_in_dim is the same as the T5 embedding dimension
+        self.context_in_dim = job_config.encoder.max_t5_encoding_len
+
+    def get_nparams_and_flops(self, model: nn.Module, seq_len: int) -> tuple[int, int]:
+        # TODO(jianiw): Add the number of flops for the autoencoder
+        logger.warning("FLUX model haven't implement get_nparams_and_flops() function")
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/model/model.py:53]: the `nparams` part should be easy, can we at least add that part?

File: torchtitan/experiments/flux/__init__.py [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+    ),
+    "flux-debug": FluxModelArgs(
+        in_channels=64,
+        out_channels=64,
+        vec_in_dim=768,
+        context_in_dim=512,
+        hidden_size=2048,
+        mlp_ratio=4.0,
+        num_heads=16,
+        depth=8,
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/__init__.py]: we can use 2 layers (the 8 for llama was because we needed to test PP)

File: torchtitan/experiments/flux/README.md [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+
+```
+
+## TODO
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/README.md:13]: since you added this TODO list, let add others too:
CI test, Tensor Parallel, Context Parallel, distributed checkpoint conversion and loading

**Review by tianyu-l — APPROVED** [review]
Awesome work! Please address remaining comments before merge.
